README.md
# Gungnir V2

[![npm](https://img.shields.io/npm/v/vuepress-theme-gungnir/next.svg?style=flat-square&logo=npm)](https://www.npmjs.com/package/vuepress-theme-gungnir/v/next) [![docs](https://img.shields.io/badge/Docs-Gungnir-26A2FF?style=flat-square)](https://v2-vuepress-theme-gungnir.vercel.app/docs/) [![license](https://img.shields.io/badge/License-Apache--2.0-green?style=flat-square)](LICENSE) [![awesome-vuepress](https://cdn.jsdelivr.net/gh/sindresorhus/awesome@a7b5b7805ecc00b3396acb35022cf6e2c0699e03/media/badge-flat.svg)](https://github.com/vuepress/awesome-vuepress/blob/main/v2.md#themes)

A blog theme for [VuePress 2](https://v2.vuepress.vuejs.org/), **WIP**. 

**Gungnir** is the spear of the god Odin in Norse mythology.

![preview](https://raw.githubusercontent.com/Renovamen/vuepress-theme-gungnir/main/docs/.vuepress/public/img/docs/gungnir.jpg)


&nbsp;

## Features

- Simple and beautiful
- Responsive design
- Multi-level navigation
- Switchable cover images for home page
- Dark and light mode: system theme detection with a theme-switcher to set the theme as per the user preference
- Search
- Table of contents
- Full screening the code blocks (inspired by WordPress theme [Sakura](https://github.com/mashirozx/Sakura))
- Tex support ([Katex](https://github.com/KaTeX/KaTeX))
- Draw charts and graphs in Markdown easily: [Chart.js](https://www.chartjs.org), [Mermaid](https://mermaid-js.github.io)
- Markdown enhancements:
  - Footnote ([markdown-it-footnote](https://github.com/markdown-it/markdown-it-footnote))
  - Mark ([markdown-it-mark](https://github.com/markdown-it/markdown-it-mark))
- Analytics ([Google Analytics](https://analytics.google.com/) and [Baidu Tongji](https://tongji.baidu.com/))
- Comment ([Giscus](https://github.com/giscus/giscus))
- RSS
- Reading time
- Multi-language support (English, Simplified Chinese)
- Archive page with tags and links page

TODO:

- [roughViz.js](https://github.com/jwilber/roughViz) and [markmap](https://github.com/gera2ld/markmap)


&nbsp;

## Installation

Please make sure you are using the latest version of VuePress (2.0.0-beta.49) and theme Gungnir.

```bash
pnpm install -D vuepress vuepress-theme-gungnir@next
```


&nbsp;

## Documentation & Live Demos

- Documentation: [v2-vuepress-theme-gungnir.vercel.app](https://v2-vuepress-theme-gungnir.vercel.app)
- My blog: [blog.zxh.io](https://blog.zxh.io)


&nbsp;

## Plugins

This repository also contains the following plugins:

- [plugin-giscus](packages/plugins/giscus): use [Giscus](https://github.com/giscus/giscus) (a comments system powered by [GitHub Discussions](https://docs.github.com/en/discussions)) in Vuepress 2
- [plugin-katex](packages/plugins/katex): [KaTeX](https://katex.org/)
- [plugin-chart](packages/plugins/chart): [Chart.js](https://www.chartjs.org)
- [plugin-mermaid](packages/plugins/mermaid): [Mermaid](https://mermaid-js.github.io)
- [plugin-reading-time](packages/plugins/reading-time): word count and reading time
- [plugin-baidu-tongji](packages/plugins/baidu-tongji): [百度统计](https://tongji.baidu.com/)
- [plugin-rss](packages/plugins/rss): RSS
- [plugin-code-enhance](packages/plugins/code-enhance): full-screen and ~~copy~~ button for code blocks
- [plugin-search](packages/plugins/search): edit [@vuepress/plugin-search](https://github.com/vuepress/vuepress-next/tree/main/packages/%40vuepress/plugin-search) to make it compatible with Gungnir


&nbsp;

## Changes (v0 -> v2)

- With the power of [VuePress 2](https://v2.vuepress.vuejs.org/), [Vue 3](https://v3.vuejs.org/), [Vite](https://vitejs.dev/) and [Webpack 5](https://webpack.js.org/), written using [Composition API](https://v3.vuejs.org/guide/composition-api-introduction.html), of course
- Stylus -> SCSS
- Drop external fonts to increase load speed
- Built-in comments system: move from [Vssue](https://github.com/meteorlxy/vssue) (doesn't support VuePress 2 yet) to [Giscus](https://github.com/giscus/giscus)


&nbsp;

## Contribution

See [Contributing Guide](CONTRIBUTING.md).


&nbsp;

## Other Versions

No further update expected for the following previous versions:

- [VuePress 1 theme](https://github.com/Renovamen/vuepress-theme-gungnir/tree/v0)
- [Jekyll theme](https://github.com/Renovamen/jekyll-theme-gungnir)


&nbsp;

## License

VuePress theme Gungnir is [Apache License 2.0](https://www.apache.org/licenses/LICENSE-2.0) licensed, see the [LICENSE](LICENSE) file for details.

Plugins are licensed by MIT.


CHANGELOG.md
# [2.0.0-alpha.26](https://github.com/Renovamen/vuepress-theme-gungnir/compare/v2.0.0-alpha.25...v2.0.0-alpha.26) (2022-08-19)


### Bug Fixes

* home page pager arrows ([9343d7e](https://github.com/Renovamen/vuepress-theme-gungnir/commit/9343d7e6a3ddbdba0b9fe033add8a8dc070544bd))
* scrollbar style in code blocks [#77](https://github.com/Renovamen/vuepress-theme-gungnir/issues/77) ([#81](https://github.com/Renovamen/vuepress-theme-gungnir/issues/81)) ([1125d08](https://github.com/Renovamen/vuepress-theme-gungnir/commit/1125d089bbdd199636ed9d4ba832b0300ec49fff))


### Performance Improvements

* replace span element with a element to optimize SEO ([#78](https://github.com/Renovamen/vuepress-theme-gungnir/issues/78)) ([ebf8638](https://github.com/Renovamen/vuepress-theme-gungnir/commit/ebf8638a4f309fd2558eb0c652be943d9ccc27ca))



# [2.0.0-alpha.25](https://github.com/Renovamen/vuepress-theme-gungnir/compare/v2.0.0-alpha.24...v2.0.0-alpha.25) (2022-06-20)


### Bug Fixes

* replaceAll -> replace, to support Node 14 [#69](https://github.com/Renovamen/vuepress-theme-gungnir/issues/69) ([766a0e5](https://github.com/Renovamen/vuepress-theme-gungnir/commit/766a0e56919b1bfcf0d7b46c28f0003704f951ba))


### Features

* support `colorMode` and `colorModeSwitch` options ([901d635](https://github.com/Renovamen/vuepress-theme-gungnir/commit/901d6355efd3126e7567bd27efd25803a2826cf4))



# [2.0.0-alpha.24](https://github.com/Renovamen/vuepress-theme-gungnir/compare/v2.0.0-alpha.23...v2.0.0-alpha.24) (2022-05-29)


### Bug Fixes

* code block line number, close [#65](https://github.com/Renovamen/vuepress-theme-gungnir/issues/65) ([d98067f](https://github.com/Renovamen/vuepress-theme-gungnir/commit/d98067f4856b4a78b1427145adbd68c30d552e18))
* compatible with vuepress v2.0.0-beta.46 [#69](https://github.com/Renovamen/vuepress-theme-gungnir/issues/69) [#68](https://github.com/Renovamen/vuepress-theme-gungnir/issues/68) [#61](https://github.com/Renovamen/vuepress-theme-gungnir/issues/61) ([f28e131](https://github.com/Renovamen/vuepress-theme-gungnir/commit/f28e131a4ca73d12eafe274df8d61da01c2ccd51))


### Features

* change fullscreen button display ([#66](https://github.com/Renovamen/vuepress-theme-gungnir/issues/66)) ([1c0b2d7](https://github.com/Renovamen/vuepress-theme-gungnir/commit/1c0b2d777d84e6f2f054c2a0de1e068a2bd30575))
* improve pager style & add sub&sup syntax suppor ([#67](https://github.com/Renovamen/vuepress-theme-gungnir/issues/67)) ([c152796](https://github.com/Renovamen/vuepress-theme-gungnir/commit/c152796655f7dc2be015141bfc1beb2615401d8d))
* support disabling Giscus of a certain page ([d874e5a](https://github.com/Renovamen/vuepress-theme-gungnir/commit/d874e5a0f1659f2d764427ab0f987b8a87156771))



# [2.0.0-alpha.23](https://github.com/Renovamen/vuepress-theme-gungnir/compare/v2.0.0-alpha.22...v2.0.0-alpha.23) (2022-05-20)


### IMPORTANT

To be compatible with VuePress >= 2.0.0-beta.40, this release contains important breaking changes, see [here](https://github.com/vuepress/vuepress-next/blob/main/CHANGELOG.md#200-beta40-2022-04-25) for details.

You could migrate your config file as follow, and check our latest docs for more details:

```diff
- module.exports = {
-   theme: "gungnir",
-   themeConfig: {
-     // your theme configs
-   }
- }

+ const { gungnirTheme } = require("vuepress-theme-gungnir");
+
+ module.exports = {
+   theme: gungnirTheme({
+     // you theme configs
+   })
+ }
```

```diff
- module.exports = {
-   plugins: [
-     [
-       "vuepress-plugin-giscus",
-       {
-         repo: "[repo]",
-         repoId: "[repo id]",
-         category: "[category name]",
-         categoryId: "[category id]"
-       }
-     ]
-   ]
- }

+ const { giscusPlugin } = require("vuepress-plugin-giscus");
+
+ module.exports = {
+   plugins: [
+     giscusPlugin({
+         repo: "[repo]",
+         repoId: "[repo id]",
+         category: "[category name]",
+         categoryId: "[category id]"
+     })
+   ]
+ }
```


### Bug Fixes

* **plugin-chart:** replace eval() with safer Function() ([ffd7913](https://github.com/Renovamen/vuepress-theme-gungnir/commit/ffd7913335971bc033e7234552065a39d72546c2))


### Features

* improve SEO (render blog list on server side) [#62](https://github.com/Renovamen/vuepress-theme-gungnir/issues/62) [#28](https://github.com/Renovamen/vuepress-theme-gungnir/issues/28) ([419b2c8](https://github.com/Renovamen/vuepress-theme-gungnir/commit/419b2c82169d2337283055ed27db2bda13710834))





# [2.0.0-alpha.22](https://github.com/Renovamen/vuepress-theme-gungnir/compare/v2.0.0-alpha.21...v2.0.0-alpha.22) (2022-05-01)


### Bug Fixes

* catalog height [#60](https://github.com/Renovamen/vuepress-theme-gungnir/issues/60) ([e3483b2](https://github.com/Renovamen/vuepress-theme-gungnir/commit/e3483b2ed3d97562cdb91e4fb71e62459d3c7aa0))





# [2.0.0-alpha.21](https://github.com/Renovamen/vuepress-theme-gungnir/compare/v2.0.0-alpha.20...v2.0.0-alpha.21) (2022-04-18)


### Bug Fixes

* **plugin-giscus:** fix giscus lang options setting, close [#57](https://github.com/Renovamen/vuepress-theme-gungnir/issues/57) ([#58](https://github.com/Renovamen/vuepress-theme-gungnir/issues/58)) ([bb200e2](https://github.com/Renovamen/vuepress-theme-gungnir/commit/bb200e230ef2179d96ea4fec7f0d41d850994e70))
* pwa default option ([a3d680a](https://github.com/Renovamen/vuepress-theme-gungnir/commit/a3d680ad28e280e5971e5b8785a9e1e4021a8307))





# [2.0.0-alpha.20](https://github.com/Renovamen/vuepress-theme-gungnir/compare/v2.0.0-alpha.19...v2.0.0-alpha.20) (2022-04-13)


### Bug Fixes

* compatible with vuepress v2.0.0-beta.39 [#55](https://github.com/Renovamen/vuepress-theme-gungnir/issues/55) ([e68d31b](https://github.com/Renovamen/vuepress-theme-gungnir/commit/e68d31b404cea050a0ca8847d99dfa4c7244498f))

example/.vuepress/config.js
module.exports = {
  title: "vuepress-theme-gungnir",
  description: "Gungnir is a blog theme for VuePress.",
  head: [
    ["link", { rel: "icon", href: "/img/logo.svg" }],
    ["meta", { name: "viewport", content: "width=device-width,initial-scale=1,user-scalable=no" }]
  ],
  theme: "gungnir",
  locales: {
    "/": {
      lang: "en-US"
    },
    "/zh/": {
      lang: "zh-CN"
    }
  },
  themeConfig: {
    repo: "Renovamen/vuepress-theme-gungnir",
    docsDir: "example",
    editLinks: true,
    lastUpdated: true,
    hitokoto: true,  // enable hitokoto (一言) or not?
    searchIcon: "ri-search-2-line",
    languageIcon: "hi-translate",
    rss: {
      site_url: "https://vuepress-theme-gungnir.vercel.app",
      copyright: "Renovamen 2018-2021",
      count: 20
    },
    comment: {
      owner: "This-is-an-Apple",
      repo: "gitalk-comments",
      clientId: "0a24dcc429074a13836a",
      clientSecret: "8d805542796c114a131012d2850a1e4d43772d98"
    },
    analytics: {
      ga: "UA-146858305-4",
      ba: "0958eaa31f4f4656f36bd33673332939"
    },
    katex: true,
    mermaid: true,
    chartjs: true,
    roughviz: true,
    markmap: true,
    mdPlus: {
      all: true
    },
    locales: {
      "/": {
        label: "English",
        selectText: "Languages",
        nav: require("./configs/nav/en"),
        sidebar: require("./configs/sidebar/en")
      },
      "/zh/": {
        label: "简体中文",
        selectText: "选择语言",
        nav: require("./configs/nav/zh"),
        sidebar: require("./configs/sidebar/zh")
      }
    },
    personalInfo: {
      name: "Gungnir",
      avatar: "/img/avatar.jpeg",
      description: "A blog theme for VuePress.",
      sns: {
        github: "Renovamen",
        linkedin: "xiaohan-zou-55bba0160",
        facebook: "renovamen.zou",
        twitter: "renovamen_zxh",
        zhihu: "chao-neng-gui-su",
        email: "renovamenzxh@gmail.com"
      }
    },
    homeHeaderImages: {
      local: [
        {
          path: "/img/home-bg/1.jpg",
          mask: "rgba(40, 57, 101, .4)"
        },
        {
          path: "/img/home-bg/2.jpg",
          mask: "rgb(251, 170, 152, .2)"
        },
        {
          path: "/img/home-bg/3.jpg",
          mask: "rgba(68, 74, 83, .1)"
        },
        {
          path: "/img/home-bg/4.jpg",
          mask: "rgba(19, 75, 50, .2)"
        }
      ]
    },
    pages: {
      tags: {
        subtitle: "Black Sheep Wall",
        bgImage: {
          path: "/img/pages/tags.jpg",
          mask: "rgba(211, 136, 37, .5)"
        }
      },
      links: {
        subtitle:
          "When you are looking at the stars, please put the brightest star shining night sky as my soul.",
        bgImage: {
          path: "/img/pages/links.jpg",
          mask: "rgba(64, 118, 190, 0.5)"
        }
      }
    },
    footer: `
      &copy; <a href="https://github.com/Renovamen" target="_blank">Renovamen</a> 2018-2021
      <br>
      Powered by <a href="https://vuepress.vuejs.org" target="_blank">VuePress</a> &
      <a href="https://github.com/Renovamen/vuepress-theme-gungnir" target="_blank">Gungnir</a>
    `
  },
  markdown: {
    // lineNumbers: true,
    extractHeaders: ["h2", "h3", "h4", "h5"]
  }
};
